model/monitorCache/implementation/MonitoringStorage.php

```diff
@@ -328,27 +320,60 @@ public function count(array $criteria = [])
 
     /**
      * @param DeliveryMonitoringDataInterface $deliveryMonitoring
-     * @return boolean whether data is saved
+     * @return bool|mixed
+     * @throws \Exception
      */
     public function save(DeliveryMonitoringDataInterface $deliveryMonitoring)
     {
         $result = false;
         if ($deliveryMonitoring->validate()) {
-            $isNewRecord = $this->isNewRecord($deliveryMonitoring);
-
-            if ($isNewRecord) {
+            try {
+                // we should be ready for unique violation error when the calling side calls
+                // save() instead of partialSave()
                 $result = $this->create($deliveryMonitoring);
-            } else {
+            } catch (\PDOException $e) {
+            } catch (DBALException $e) {
+            }
+            if (!$result) {
                 $result = $this->update($deliveryMonitoring);
+                $this->saveKvData($deliveryMonitoring);
             }
         }
         return $result;
     }
 
+    /**
+     * @param DeliveryMonitoringDataInterface $deliveryMonitoring
+     * @return bool|mixed
+     * @throws \Exception
+     */
+    public function partialSave(DeliveryMonitoringDataInterface $deliveryMonitoring)
+    {
+        $result = false;
+        if ($deliveryMonitoring->validate()) {
+            $rowsUpdated = $this->update($deliveryMonitoring);
+            if ($rowsUpdated === 0) {
+                // doesn't mean an error for sure, cause persistence may return the number of rows actually changed,
+                // and not the number of rows matched by the where clause.
+                // So just in case try to create without fallback
+                try {
+                    $this->create($deliveryMonitoring);
+                } catch (\PDOException $e) {
+                } catch (DBALException $e) {
+                }
+            }
+            $this->saveKvData($deliveryMonitoring);
+            $result = true;
```

Comment:
`$result` will be `true` even in case of `PDOException` and `DBALException`.

Reply:
the saveKvData() should probably be part of an else block, because the create includes the storing of keyvalue data

Reply:
@hutnikau that's intended, even when there is an exception, it means that the rows already exist and we just couldn't reveal them with update, as it returned 0 actually changed rows

Reply:
If `create` call fails then `saveKvData` will fail too, so it does not make sense to catch those exceptions as it will definitely fail on next line of code. Maybe it makes sense to wrap `saveKvData` in `try` as well?

Reply:
@jbout  The save/saveKv logic has been fixed
@hutnikau  Now the UniqueConstraintViolationException is used, so any general error will prevent executing saveKvData (that doesn't work if you use PDO instead of DBAL, because PDO has only the broad exception, but it's being fixed as per TAO-7005).